Implemented CLI upload functionality (#1618)
huggingface/huggingface_hub: docs/source/en/guides/upload.md

... )
```

## Upload from the CLI

> You can also upload files to the Hub directly from your terminal using the `huggingface-cli upload` command. Internally

Reviewer: ```suggestion
You can use the `huggingface-cli upload` command from the terminal to directly upload files to the Hub. Internally
```

it uses the same [`upload_file`] and [`upload_folder`] helpers described above.

You can either upload a single file or an entire folder:

```bash
# Usage:  huggingface-cli upload [repo_id] [local_path] [path_in_repo]
>>> huggingface-cli upload Wauplin/my-cool-model ./models/model.safetensors model.safetensors
https://huggingface.co/Wauplin/my-cool-model/blob/main/model.safetensors

>>> huggingface-cli upload Wauplin/my-cool-model ./models .
https://huggingface.co/Wauplin/my-cool-model/tree/main
```

> `local_path` and `path_in_repo` are optional and can be implicitly inferred. By default, `local_path` will be set to
> the current directory and `path_in_repo` will be set to the relative path between the current directory and `local_path`.

Reviewer: ```suggestion
`local_path` and `path_in_repo` are optional and can be implicitly inferred. By default, `local_path` is set to
the current directory, and `path_in_repo` is set to the relative path between the current directory and `local_path`.
```

If the implicit paths cannot be inferred, an error is raised.

```bash
# Upload file (implicit path_in_repo)
huggingface-cli upload my-cool-model model.safetensors

# Upload directory (implicit path_in_repo)

> huggingface-cli upload my-cool-model ./models

Reviewer 2: Updated the CLI and now that's the case. Also added a unit test for this use case.

# Upload directory (implicit local_path, implicit path_in_repo)

> huggingface-cli upload my-cool-model

Reviewer 2: Thanks for the feedback @LysandreJik. My initial thought was that `huggingface-cli upload my-cool-model` would upload the current directory to the Hub (at root level). So `huggingface-cli upload my-cool-model` being equivalent to `huggingface-cli upload my-cool-model . .` . A bit as when you have a local repo and you do `git add . && git commit -m "something" && git push`. 

That being said, the use case you've just described (e.g. repo_id == name of a local folder) is perfectly valid as well. I've updated the CLI in that sense. Now the behavior looks like this:

```bash
# Current repo structure
# ./
# ../
# folder-1/
#     file.txt
# folder-2/
#     file.txt

# Upload "./folder-1" content to "Wauplin/folder-1"
# On the hub: file.txt
>>> huggingface-cli upload folder-1

# Upload "./folder-1" content to "huggingface/folder-1"
# On the hub: file.txt
>>> huggingface-cli upload huggingface/folder-1

# Upload "./folder-1" and "./folder-2" to "Wauplin/my-cool-model"
# On the hub: folder-1/file.txt and folder-2/file.txt
>>> huggingface-cli upload my-cool-model .

# Upload "./folder-1" and "./folder-2" to "Wauplin/my-cool-model" under "./data"
# On the hub: data/folder-1/file.txt and data/folder-2/file.txt
>>> huggingface-cli upload my-cool-model . data/

# Raise exception => user must set local path explicitly 
>>> huggingface-cli upload folder-3
```
